FWCore/Framework/interface/CallbackExternalWork.h

```diff
@@ -194,6 +198,7 @@ namespace edm {
       }
 
       std::shared_ptr<TAcquireFunc> acquireFunction_;
+      TAcquireReturn acquireCache_;
```

Comment:
After staring at this for a while, I gave it a try. I didn't understand the discussion completely when we talked about this. It is only a two line change. Looks trivial. It seems to me the easiest way to do this. It has the slightly odd feature that the data member will be a ```std::optional<std::optional<X>>``` in the case where acquire returns a ```std::optional<X>```. Although that does not seem a problem to me. Can you take a look? It is the second commit. I'll be pushing more commits later. I intentionally didn't rerun all the tests yet, but the only test configuration affected by this commit passes in my working area.

Reply:
It is a starting point. I wonder if `acquireCache_.reset()` should be called in the lambda constructed in `CallbackExternalWork::prefetch()`. In principle it would be cleaner to leave the `optional` in an without-value state than to keep moved-from value in the `optional`, but the practical impact seems minimal.

> It has the slightly odd feature that the data member will be a `std::optional<std::optional<X>>` in the case where acquire returns a `std::optional<X>`

This is exactly about the "further optimization" that I mentioned above. Nesting `std::optional` comes with a small memory cost, and in principle it would be fairly straightforward to avoid. It could be e.g. along
```cpp
template <typename T>
struct AcquireCacheType {
  using type = std::optional<T>;
};
template <T>
struct AcquireCacheType<std::optional<T>> {
  using type = std::optional<T>;
};
template <T>
struct AcquireCacheType<std::unique_ptr<T>> {
  using type = std::unique_ptr<T>;
};
template <T>
struct AcquireCacheType<std::shared_ptr<T>> {
  using type = std::shared_ptr<T>;
};
...
typename AcquireCacheType<TAcquireReturn>::type acquireCache_;
```
(I'm sure other implementations could be possible as well)

Reply:
I used that. I think this is done now. Thanks.